### PR TITLE
SpectralCube to specutils transition in cubeviz

### LINKED ISSUE #544: Update cubeviz to work with latest Spectrum1D
This ticket tracks the work needed to replace SpectralCube in cubeviz with the latest Spectrum1D, which can now handle cube data.

### PR DESCRIPTION
**Pull request by javerbukh**
Resolves #544

Requires [Specutils 811](https://github.com/astropy/specutils/pull/811)

Also requires a fix to [glue-astronomy 28](https://github.com/glue-viz/glue-astronomy/issues/28)

### CODE REVIEW

**Comment by javerbukh** [comment]
Latest changes can be tested by running the CubevizExample notebook to the 4th cell. Then, go to `Viewer` and change the x_axis to Declination and y_axis to Right Ascension. This change needs to be done because the latest specutils updates have a different shape from SpectralCube. I'm not sure yet how to do this programmatically and may need some assistance from @rosteen and @ibusko on how to implement this. Also, the slider seems to snap back to 0 after dragging, not sure why this is happening either.

**Comment by ibusko** [comment]
I am seeing this output, but no data gets loaded:
```
FLUX
Wrong number of dimensions in input array.  Expected 2.
IVAR
Wrong number of dimensions in input array.  Expected 2.
MASK
Wrong number of dimensions in input array.  Expected 2.
WAVE
SPECRES
SPECRESD

**Comment by javerbukh** [comment]
That is correct about `_new_parse_hdu` but the 4th cell should manually load data into `data_collection` and it should be available in the data drop down for a viewer.

**Comment by javerbukh** [comment]
Latest changes have a manga cube load into cubeviz using Spectrum1D. Some hacky things had to be done for this to work, like using `np.moveaxis()` on some axes in `hdu.data` and using Spectrum1D to load individual hdu's rather than the entire data cube at once. Also, if you check the viewer tab, you can see that the `x axis` is WAVE and the `y axis` is Declination. Not sure why this is happening but using `swapaxes` on the wcs does not change this either. I am looking for some feedback and advice on the previously mentioned issues.

**Review by duytnguyendtn — COMMENTED** [review]
Here's what I get when I change the axes:
image

which is probably not what it's supposed to look like. I spent a few hours trying different things to no avail. My only possible theory is whether us moving the axes around (to accommodate for the spectral axis) is either confusing us, or the glue backend) and causing the axes to be mislabeled. I was going to try and see if I could plot it in matplotlib to see if *I* understood the axes correctly, but promptly got confused...

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 1]
```diff
@@ -68,49 +71,34 @@ def _parse_hdu(app, hdulist, file_name=None):
...
-        # This is supposed to fail on attempting to load anything that
-        # isn't cube-shaped. But it's not terribly reliable
+        wcs = WCS(hdu.header)
+
+        if 'BUNIT' in hdu.header:
+            flux = hdu.data * u.Unit(hdu.header['BUNIT'])
+        else:
+            # TODO Add warning that actual units were not found and using fake ones
+            logging.warn("No flux units found in hdu, using Jansky as a stand-in")
+            flux = hdu.data * u.Jy
```
[the remark this answers is not in the document]
Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/parsers.py:90]: I agree with @duytnguyendtn because `u.Unit(hdu.header['BUNIT'])` is going to break for most of the FITS images you have. For example: `'ELECTRONS/S'` does not resolve to `u.electron/u.s` (`astropy.units` is case-sensitive and very pedantic about the "s"es).

However, falling back to `u.Jy` is also a problem. For example, when parsing of `'ELECTRONS/S'` fails, you get Jy, which is wrong.

[the remark this answers is not in the document]
Comment (javerbukh) [reply jdaviz/configs/cubeviz/plugins/parsers.py:90]: To start with the axes, moving the axes using `np.moveaxis(flux, 1, 2)` or `wcs = wcs.swapaxes(0, 1)` does NOT change the x or y axis labels in the glue jupyter viewer. Those labels are set using the `ComponentIDs`, I believe in glue astronomy, but hours investigating this did yield useful results. We may need some SME help on this matter.

In regards to the flux units, thank you for your suggestions Duy! I will implement your suggestions in the next commit. To answer Pey Lian, I agree that giving the user other units is a problem but I think the choice is either that or not load data at all. Although, we were having discussions previously about using `counts` as units, is that a better replacement?

[the remark this answers is not in the document]
Comment (duytnguyendtn) [reply jdaviz/configs/cubeviz/plugins/parsers.py]: I think this may be a symptom of the broader issue @aaastorrs brought up in #566. I'd argue you're right, @javerbukh; I don't think we shouldn't have to rely on assigning an arbitrary `u.Jy` when things don't work (as @pllim brings up, this can very often create a graph that is lying to the user), rather I'd argue that a generic `u.count` is at least a "correct" fallback since it is so generic if I understand the unit right

[the remark this answers is not in the document]
Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/parsers.py]: Yes, we use "count" for anything here. 😆

[the remark this answers is not in the document]
Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/parsers.py]: DzhyYE8d_400x400

**Comment by javerbukh** [comment]
TL;DR: App-wide conversion to Spectrum1D will take longer if wavelength axis is forced to be last, and the time it takes for data to load is significantly longer. Do we still want to continue development in this direction with that in mind? Is there any chance we can change the decision on the wavelength axis being last in Spectrum1D? @eteq 

The latest commit has working axes (check `_fix_axes()`) and uses `u.count` if no flux units are found, so that solves two of the concerns with this PR. The next concern would be getting the cubeviz plugins to work (i.e. use Spectrum1D instead of SpectralCube) with these changes to the parser, which is tracked [here](https://github.com/spacetelescope/jdaviz/issues/587). Since Spectrum1D forces a different order of axes from SpectralCube, we need to make sure that every place that uses a particular axis for wavelength, RA, or Dec now instead either figures out programmatically which is correct or now uses the Spectrum1D location for axes, which make take some time.

Another thought is that there may be work in glue-astronomy that can make this conversion simpler, but that development work is made more complicated by the "wavelength as last axis" decision, in my opinion. That said, I will proceed in this direction for now.

Screen Shot 2021-05-04 at 4 50 27 PM

File: notebooks/CubevizExample.ipynb [hunk 2]
```diff
@@ -27,7 +27,7 @@
     "# This file is originally from https://data.sdss.org/sas/dr14/manga/spectro/redux/v2_1_2/7495/stack/manga-7495-12704-LOGCUBE.fits.gz\n",
     "# but has been modified to correct some inconsistencies in the way units are parsed\n",
     "fn = download_file('https://stsci.box.com/shared/static/28a88k1qfipo4yxc4p4d40v4axtlal8y.fits', cache=True)\n",
-    "viz.load_data(fn)"
+    "viz.load_data(fn)\n"
```
Comment (pllim) [inline notebooks/CubevizExample.ipynb:30]: FYI -- I am going to drop the changes here in my rebase. So the cleaned up PR won't show anymore changes to the example notebook, just `parsers.py`.

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 3]
```diff
@@ -56,6 +58,8 @@ def parse_data(app, file_obj, data_type=None, data_label=None):
     #  into something glue can understand.
     elif isinstance(file_obj, SpectralCube):
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:59]: @javerbukh , shouldn't we remove this too?

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 4]
```diff
@@ -67,49 +71,34 @@ def _parse_hdu(app, hdulist, file_name=None):
...
-            wcs = sc.wcs
-
     # Now loop through and attempt to parse the fits extensions as spectral
     #  cube object. If the wcs fails to parse in any case, use the wcs
     #  information we scraped above.
     for hdu in hdulist:
         data_label = f"{file_name}[{hdu.name}]"
 
-        if hdu.data is None or not hdu.is_image:
+        if hdu.data is None or not hdu.is_image or len(hdu.data.shape) != 3:
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:80]: I am going to use `hdu.data.ndim` instead of `len(hdu.data.shape)`. FYI.

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 5]
```diff
@@ -67,49 +71,34 @@ def _parse_hdu(app, hdulist, file_name=None):
...
             continue
 
-        # This is supposed to fail on attempting to load anything that
-        # isn't cube-shaped. But it's not terribly reliable
+        wcs = WCS(hdu.header)
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:83]: You also have to pass in the hdulist to WCS to support WCS-TAB etc. I'll do that in my rebase. FYI.

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 6]
```diff
@@ -67,49 +71,34 @@ def _parse_hdu(app, hdulist, file_name=None):
...
+
+        try:
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:85]: There is also the case of BUNIT exists but invalid. I'll add logic to handle that in my rebase. FYI.

File: jdaviz/configs/cubeviz/plugins/parsers.py (hunk 5 above)
Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/parsers.py:83]: Also, I'll add a try/except here to handle cubes without valid WCS, though consequences of generating a cube with `wcs=None` remains mysterious.

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 7]
```diff
@@ -141,6 +147,36 @@ def _parse_spectral_cube(app, file_obj, data_type='flux', data_label=None):
...
 
+def _parse_spectrum1d_3d(app, file_obj):
+    # Load spectrum1d as a cube
+
+    for attr in ["flux", "mask", "uncertainty"]:
+        if attr == "mask":
+            flux = getattr(file_obj, attr) * file_obj.flux.unit
+        elif attr == "uncertainty":
+            flux = getattr(file_obj, attr)
+            flux = u.Quantity(flux.array) * file_obj.flux.unit
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:158]: I think this should be `u.Quantity(flux.array, file_obj.flux.unit)`. If `flux.array` already has unit, you are multiplying `file_obj.flux.unit` to it instead of recasting.

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 8]
```diff
@@ -141,6 +147,36 @@ def _parse_spectral_cube(app, file_obj, data_type='flux', data_label=None):
     # TODO: SpectralCube does not store data quality information
 
 
+def _parse_spectrum1d_3d(app, file_obj):
+    # Load spectrum1d as a cube
+
+    for attr in ["flux", "mask", "uncertainty"]:
+        if attr == "mask":
+            flux = getattr(file_obj, attr) * file_obj.flux.unit
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:155]: I am going to use the `<<` operator here, see https://docs.astropy.org/en/latest/units/index.html#performance-tips .

File: jdaviz/configs/cubeviz/plugins/parsers.py (hunk 7 above)
Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/parsers.py:158]: If performance is a problem, also consider whether you can get away with `copy=False` or `subok=True`. See https://docs.astropy.org/en/latest/api/astropy.units.Quantity.html#astropy.units.Quantity

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 9]
```diff
@@ -141,6 +147,36 @@ def _parse_spectral_cube(app, file_obj, data_type='flux', data_label=None):
...
+
+        flux = np.moveaxis(flux, 1, 0)
+
+        s1d = Spectrum1D(flux=flux, wcs=file_obj.wcs)
+
+        data_label = f"Unknown spectrum object[{attr}]"
+        app.data_collection[data_label] = s1d
+
+        if attr == 'flux':
+            app.add_data_to_viewer('flux-viewer', f"{data_label}")
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:170]: Why is the `f""` necessary for the `data_label`?

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 10]
```diff
@@ -141,6 +147,36 @@ def _parse_spectral_cube(app, file_obj, data_type='flux', data_label=None):
...
+            app.add_data_to_viewer('spectrum-viewer', f"{data_label}")
+        elif attr == 'mask':
+            app.add_data_to_viewer('mask-viewer', f"{data_label}")
+        elif attr == 'uncertainty':
+            app.add_data_to_viewer('uncert-viewer', f"{data_label}")
+
+    # _fix_axes(app)
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:177]: Removing unused code too, as per tag up discussions.

File: CHANGES.rst [hunk 11]
```diff
@@ -49,6 +49,9 @@ Specviz
 Other Changes and Additions
 ---------------------------
 
+- ``jdaviz`` no longer uses ``spectral-cube``. Cubeviz now loads data cube
```
Comment (pllim) [inline CHANGES.rst:52]: This is not quite true yet, but hopefully will be true by the time this PR is complete.

**Comment by pllim** [comment]
@javerbukh , I might have gotten a little carried away with the rebasing. Please double check the diff. I don't grok the flux/mask/uncert stuff but hopefully I didn't mess things up too much.

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 12]
```diff
@@ -54,10 +54,13 @@ def parse_data(app, file_obj, data_type=None, data_label=None):
     # If the data types are custom data objects, use explicit parsers. Note
     #  that this relies on the glue-astronomy machinery to turn the data object
     #  into something glue can understand.
-    elif isinstance(file_obj, SpectralCube):
-        _parse_spectral_cube(app, file_obj, data_type or 'flux', data_label)
     elif isinstance(file_obj, Spectrum1D):
-        _parse_spectrum1d(app, file_obj)
+        if file_obj.flux.ndim == 3:
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:58]: For a hacky way to load @havok2063 s3d files, see javerbukh/jdaviz#9

Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/parsers.py:58]: For the desired way to load (but unsuccessful), see javerbukh/jdaviz#10

File: setup.cfg [hunk 13]
```diff
@@ -78,7 +78,7 @@ testpaths = "jdaviz" "docs"
 astropy_header = true
 doctest_plus = enabled
 text_file_format = rst
-addopts = --doctest-rst --import-mode=append
+addopts = --doctest-rst --import-mode=append -p no:warnings
```
Comment (javerbukh) [inline setup.cfg:81]: @pllim This is what I had to do to prevent `Spectrum1D reshaping flux array` warnings from running tests in CI. Please let me know if there is an alternative to this since it feels wrong!

File: setup.cfg [hunk 14]
```diff
@@ -91,6 +91,7 @@ filterwarnings =
     ignore::DeprecationWarning:bqplot_image_gl
     ignore::DeprecationWarning:bqscales
     ignore::DeprecationWarning:traittypes
+    ignore::Warning:specutils.spectra.spectrum1d
```
Comment (pllim) [inline setup.cfg:94]: Everything subclasses `Warning` (except deprecation, I think), so this is a little redundant.
```suggestion
    ignore:::specutils.spectra.spectrum1d
```

File: CHANGES.rst [hunk 15]
```diff
@@ -62,6 +62,9 @@ Specviz
...
+  as ``Spectrum1D``. [#547]
+- MIRI s2d files can now be loaded into Specviz2d. [#915]
```
[the remark this answers is not in the document]
Comment (pllim) [reply CHANGES.rst:67]: Maybe a rebase will fix this. 🤷

File: jdaviz/configs/default/plugins/collapse/collapse.py [hunk 16]
```diff
@@ -122,6 +123,10 @@ def vue_collapse(self, *args, **kwargs):
             spec_max = float(self.spectral_max) * u.Unit(self.spectral_unit)
             spec = spec.spectral_slab(spec_min, spec_max)
 
+            # Code to work with Spectrum1D
+            # spec = manipulation.spectral_slab(spec, spec_min, spec)
+            # spec = spec.spectral_slab(spec_min, spec_max)
```
Comment (pllim) [inline jdaviz/configs/default/plugins/collapse/collapse.py:128]: Do we need to keep this?

**Review by pllim — COMMENTED** [review]
p.s. I am basically a co-author by now, so I'll abstain from approving.
